File: internal/actions/email.go
package actions

import (
	"context"
	"fmt"
	"net/mail"
	"strings"

	"github.com/openforms/openforms/internal/jobs"
)

func (r *runner) email(ctx context.Context, job jobs.Job) error {
	ac, err := r.load(ctx, job)
	if err != nil {
		return err
	}
	a := ac.payload.Action
	to := strings.TrimSpace(Render(a.To, ac.vars))
	if to == "" {
		return jobs.Permanent(fmt.Errorf("email: recipient template %q rendered empty", a.To))
	}
	if _, err := mail.ParseAddress(to); err != nil || strings.ContainsAny(to, "\r\n") {
		return jobs.Permanent(fmt.Errorf("email: invalid recipient %q", to))
	}
	if err := r.d.Mailer.Send(ctx, to, Render(a.Subject, ac.vars), Render(a.Body, ac.vars)); err != nil {
		return fmt.Errorf("email: %w", err)
	}
	return r.recordSuccess(ctx, ac)
}

File: src/openforms/server/actions/email.py
"""The email action: render recipient/subject/body and send."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ...definition.common import is_bare_email
from ..services.jobs import Job, Permanent
from .render import render

if TYPE_CHECKING:
    from .runner import Runner


async def run_email(r: Runner, job: Job) -> None:
    ac = await r.load(job)
    a = ac.payload.action
    to = render(a.to, ac.vars).strip()
    if not to:
        raise Permanent(ValueError(f"email: recipient template {a.to!r} rendered empty"))
    if "\r" in to or "\n" in to or not is_bare_email(to):
        raise Permanent(ValueError(f"email: invalid recipient {to!r}"))
    try:
        await r.mailer.send(to, render(a.subject, ac.vars), render(a.body, ac.vars))
    except Exception as e:
        raise RuntimeError(f"email: {e}") from e
    await r.record_success(ac)
